File: lectrick/tiles/press_detector.py
from .base import Tile


class PressDetector(Tile):
    TILE_TYPE = "PRESS DETECTOR"
    SAMPLE_CHARS = ["␣"]

    def __init__(self, x, y, tiles):
        super().__init__(x, y, tiles)

    def start_turn(self, execution_data, *args, **kwargs):
        pressed_key = execution_data.get_pressed_key()
        if pressed_key is not None:
            for dx in range(-1, 2):
                for dy in range(-1, 2):
                    if (dx == 0 and dy == 0) or (dx != 0 and dy != 0):
                        continue
                    target_x = self.x + dx
                    target_y = self.y + dy
                    try:
                        tile_index = self.tiles.index((target_x, target_y))
                    except ValueError:
                        continue
                    self.tiles[tile_index].accept_energy(pressed_key, self.x, self.y)

File: web/src/tiles/pressDetector.ts
import {Tile} from './base';

export class PressDetector extends Tile {
    TILE_TYPE = 'PRESS DETECTOR';
    SAMPLE_CHARS = ["␣"];

    constructor(ctx: ExecutionContext, x: number, y: number) {
        super(ctx, x, y);
    }

    startTurn() {
        const pressedKey = this.ctx.getPressedKey();
        if (pressedKey !== null) {
            for (let dx = 0; dx < 2; dx++) {
                for (let dy = 0; dy < 2; dy++) {
                    if ((dx == 0 && dy == 0) || (dx != 0 && dy != 0)) {
                        continue;
                    }
                    const targetX = this.x + dx;
                    const targetY = this.y + dy;
                    const targetTile = this.ctx.getTile(targetX, targetY);
                    if (targetTile) {
                        targetTile.acceptEnergy(
                            pressedKey.charCodeAt(0),
                            this.x, this.y
                        );
                    }
                }
            }
        }
    }
}